src/Mvc/Mvc.Core/src/ModelBinding/Binders/DateOnlyModelBinder.cs

```diff
@@ -0,0 +1,104 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Globalization;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Binders;
+
+/// <summary>
+/// An <see cref="IModelBinder"/> for <see cref="DateOnly"/> and nullable <see cref="DateOnly"/> models.
+/// </summary>
+public class DateOnlyModelBinder : IModelBinder
+{
+    private readonly DateTimeStyles _supportedStyles;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="DateOnlyModelBinder"/>.
+    /// </summary>
+    /// <param name="supportedStyles">The <see cref="DateTimeStyles"/>.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+    public DateOnlyModelBinder(DateTimeStyles supportedStyles, ILoggerFactory loggerFactory)
+    {
+        if (loggerFactory == null)
+        {
+            throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        _supportedStyles = supportedStyles;
+        _logger = loggerFactory.CreateLogger<DateOnlyModelBinder>();
+    }
+
+    /// <inheritdoc />
+    public Task BindModelAsync(ModelBindingContext bindingContext)
+    {
+        if (bindingContext == null)
+        {
+            throw new ArgumentNullException(nameof(bindingContext));
+        }
+
+        _logger.AttemptingToBindModel(bindingContext);
+
+        var modelName = bindingContext.ModelName;
+        var valueProviderResult = bindingContext.ValueProvider.GetValue(modelName);
+        if (valueProviderResult == ValueProviderResult.None)
+        {
+            _logger.FoundNoValueInRequest(bindingContext);
+
+            // no entry
+            _logger.DoneAttemptingToBindModel(bindingContext);
+            return Task.CompletedTask;
+        }
+
+        var modelState = bindingContext.ModelState;
+        modelState.SetModelValue(modelName, valueProviderResult);
+
+        var metadata = bindingContext.ModelMetadata;
+        var type = metadata.UnderlyingOrModelType;
+        try
+        {
+            var value = valueProviderResult.FirstValue;
+
+            object? model;
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                // Parse() method trims the value (with common DateTimeSyles) then throws if the result is empty.
+                model = null;
+            }
+            else if (type == typeof(DateOnly))
+            {
+                model = DateOnly.Parse(value, valueProviderResult.Culture, _supportedStyles);
+            }
+            else
+            {
+                throw new NotSupportedException();
```

Comment:
Not sure what the general pattern is for model binding, but is this exception needed? Looks like it's immediately caught within the same method, so it's somewhat unusual to use exceptions for flow control within a method.

Reply:
I believe the intention for this is that since it's a public class, catching this exception prevents a customer from potentially creating a DIY ModelBinderProvider and feeding a non-null wrong type into an existing model binder. This implementation was carried over from the `DateTimeModelBinder`, so I was reluctant to remove it.
https://github.com/dotnet/aspnetcore/blob/f52ae1f2e40b78fdb84c4022af04e157cf503c38/src/Mvc/Mvc.Core/src/ModelBinding/Binders/DateTimeModelBinder.cs#L75-L78